# Atividade 2 — Exercícios de Limpeza, Transformação e Agrupamento de Dados

Responda as perguntas abaixo usando o dataset **emendas_tratamento.parquet**. o dataset pode ser baixado em https://drive.google.com/file/d/19ydA23-eboyiKxN29JL1EQWW4CplTiKg/view?usp=sharing.

Cada resposta deve ter:
- Uma célula de **código** com a solução
- Uma célula de **texto** (Markdown) com a sua interpretação do resultado

---
## Passo 0 — Importar pandas e carregar os dados

In [120]:
import pandas as pd 

dados = pd.read_parquet("https://drive.google.com/uc?export=download&id=19ydA23-eboyiKxN29JL1EQWW4CplTiKg")

display(dados)
dados.info()


,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
0,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ALVORADA - RS,4300604,ALVORADA,4300000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,700000.00,0.0,0.0,0.0,690900.00,9100.00,Sem informação
1,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTA ISABEL DO PARÁ - PA,1506500,SANTA ISABEL DO PARÁ,1500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,487500.00,0.0,0.0,0.0,487500.00,0.00,Sem informação
2,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,BARRA BONITA - SC,4202099,BARRA BONITA,4200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,100000.00,0.0,0.0,0.0,0.00,100000.00,Sem informação
3,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,ARARAQUARA - SP,3503208,ARARAQUARA,3500000,...,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA,0,APOIO A PROJETOS DE INFRAESTRUTURA TURISTICA -...,243750.00,0.0,0.0,0.0,243750.00,0.00,Sem informação
4,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,CAMPO LARGO - PR,4104204,CAMPO LARGO,4100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,1658333.94,0.0,0.0,0.0,527200.00,1131133.94,Sem informação
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
88654,201811420017,2018,Emenda Individual - Transferências com Finalid...,1142.0,ZE GERALDO,17,PARÁ (UF),Sem informação,Sem informação,1500000,...,CONSTRUCAO E AMPLIACAO DE UNIDADES OPERACIONAI...,EIND,EMENDA INDIVIDUAL,500000.00,0.0,0.0,0.0,369387.36,130612.64,201811420017
88655,201827690001,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,1,MÚLTIPLO,Sem informação,Múltiplo,3100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,EIND,EMENDA INDIVIDUAL,499975.00,499975.0,499975.0,0.0,0.00,0.00,201827690001
88656,201827690002,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,2,ITURAMA - MG,3134400,ITURAMA,3100000,...,APOIO A PROJETOS DE DESENVOLVIMENTO SUSTENTAVE...,EIND,EMENDA INDIVIDUAL,200000.00,0.0,0.0,11058.0,0.00,188942.00,201827690002
88657,201827690003,2018,Emenda Individual - Transferências com Finalid...,2769.0,ZE SILVA,3,ITURAMA - MG,3134400,ITURAMA,3100000,...,ESTRUTURACAO DE UNIDADES DE ATENCAO ESPECIALIZ...,EIND,EMENDA INDIVIDUAL,199970.00,199970.0,199970.0,0.0,0.00,0.00,201827690003


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88659 entries, 0 to 88658
Data columns (total 29 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Código da Emenda                    88659 non-null  object 
 1   Ano da Emenda                       88659 non-null  int64  
 2   Tipo de Emenda                      88659 non-null  object 
 3   Código do Autor da Emenda           72698 non-null  float64
 4   Nome do Autor da Emenda             72698 non-null  object 
 5   Número da emenda                    88659 non-null  object 
 6   Localidade de aplicação do recurso  88659 non-null  object 
 7   Código Município IBGE               88659 non-null  object 
 8   Município                           88659 non-null  object 
 9   Código UF IBGE                      88659 non-null  int64  
 10  UF                                  88659 non-null  object 
 11  Região                              88659

## 1) Filtre apenas as emendas do ano de **2022** e mostre as 30 primeiras linhas. Em seguida responda: Quanta emendas foram registradas em 2022?

In [121]:
dados_2022 = dados[dados["Ano da Emenda"] == 2022]

display(dados_2022.head(30))
print(len(dados_2022))

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
63473,202238950001,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,1,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950001
63474,202238950002,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,2,CONCEIÇÃO DO ALMEIDA - BA,2908309,CONCEIÇÃO DO ALMEIDA,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,750141.00,750141.00,750141.00,0.00,0.0,0.00,202238950002
63475,202238950003,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,3,CRUZ DAS ALMAS - BA,2909802,CRUZ DAS ALMAS,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950003
63476,202238950004,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,4,IRAMAIA - BA,2914307,IRAMAIA,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950004
63477,202238950005,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,5,MUCURI - BA,2922003,MUCURI,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,500000.00,0.00,0.00,500000.00,0.0,500000.00,202238950005
63478,202238950006,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,6,SERRA DO RAMALHO - BA,2930154,SERRA DO RAMALHO,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,535815.00,535815.00,535815.00,0.00,0.0,0.00,202238950006
63479,202238950007,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,7,SOUTO SOARES - BA,2930808,SOUTO SOARES,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,200000.00,0.00,0.00,200000.00,0.0,200000.00,202238950007
63480,202238950008,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,8,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,470837.00,470837.00,470837.00,0.00,0.0,0.00,202238950008
63481,202238950009,2022,Emenda Individual - Transferências Especiais,3895.0,ABILIO SANTANA,9,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,TRANSFERENCIAS ESPECIAIS,0,TRANSFERENCIAS ESPECIAIS,1100000.00,0.00,0.00,1100000.00,0.0,1100000.00,202238950009
63482,202238950010,2022,Emenda Individual - Transferências com Finalid...,3895.0,ABILIO SANTANA,10,MÚLTIPLO,Sem informação,Múltiplo,2900000,...,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,0,INCREMENTO TEMPORARIO AO CUSTEIO DOS SERVICOS ...,8573048.00,8123048.00,8123048.00,450000.00,0.0,450000.00,202238950010


6108


Foram registradas 6108 emendas no ano de 2022

## 2) Filtre as emendas da região **Norte** que tenham `Valor Empenhado` maior que **500.000** e mostre as 5 primeiras linhas. Em seguida, responda: quantas emendas atendem a esses duas condições ao mesmo tempo?

In [122]:
display(dados["Região"].unique().tolist())
dados_norte_rico = dados[(dados["Região"] == "Norte") & (dados["Valor Empenhado"] > 500000)]

display(dados_norte_rico.head())
print(len(dados_norte_rico))


['Sul',
 'Norte',
 'Sudeste',
 'Nordeste',
 'Centro-Oeste',
 'Nacional',
 'Múltiplo',
 'Exterior']

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
18,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,MÂNCIO LIMA - AC,1200336,MÂNCIO LIMA,1200000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,616610.0,0.0,0.0,0.0,0.0,616610.0,Sem informação
120,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,VITÓRIA DO JARI - AP,1600808,VITÓRIA DO JARI,1600000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,522384.4,0.0,0.0,0.0,0.0,522384.4,Sem informação
160,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SANTANA - AP,1600600,SANTANA,1600000,...,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,0,IMPLEMENTACAO DE INFRAESTRUTURA BASICA NOS MUN...,4443000.0,0.0,0.0,0.0,4443000.0,0.0,Sem informação
170,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,PALMEIRAS DO TOCANTINS - TO,1713809,PALMEIRAS DO TOCANTINS,1700000,...,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,0,APOIO A POLITICA NACIONAL DE DESENVOLVIMENTO U...,935350.0,0.0,0.0,0.0,0.0,935350.0,Sem informação
218,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,COSTA MARQUES - RO,1100080,COSTA MARQUES,1100000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,549995.0,0.0,0.0,0.0,159996.0,389999.0,Sem informação


6230


Há 6230 emendas do norte com valor superior a 500 mil

## 3) Filtre as emendas que são do estado de **SÃO PAULO** ou do estado do **RIO DE JANEIRO** e mostre quantas linhas cada estado possui. Em seguida, responda: Qual dos dois estados possui mais emendas e quantas são?

In [123]:
display(dados["Município"].unique().tolist())
dados_sp_rj = dados[(dados["Município"] == "SÃO PAULO") | (dados["Município"] == "RIO DE JANEIRO")]
display(dados_sp_rj.head())
print(f"sp:{len(dados[dados["Município"] == "SÃO PAULO"])}")
print(f"rj:{len(dados[dados["Município"] == "RIO DE JANEIRO"])}")

['ALVORADA',
 'SANTA ISABEL DO PARÁ',
 'BARRA BONITA',
 'ARARAQUARA',
 'CAMPO LARGO',
 'CAMPINAS DO SUL',
 'MORENO',
 'NOVA FLORESTA',
 'SANTIAGO DO SUL',
 'CIANORTE',
 'IPORÁ',
 'BRUMADO',
 'GUARAMIRANGA',
 'CRUZÍLIA',
 'SÃO MATEUS DO MARANHÃO',
 'SÃO GONÇALO DO AMARANTE',
 'ITALVA',
 'RIACHÃO DO JACUÍPE',
 'MÂNCIO LIMA',
 'PORTO MURTINHO',
 'PILAR DO SUL',
 'AMONTADA',
 'COMODORO',
 'SOBRADINHO',
 'BOM JESUS',
 'MONTE BELO DO SUL',
 'CABO DE SANTO AGOSTINHO',
 'PICUÍ',
 'ITAPECURU MIRIM',
 'IBIRAPUITÃ',
 'VALENÇA',
 'PORTO WALTER',
 'CASTANHAL',
 'SANTO ANTÔNIO DO PALMA',
 'APUCARANA',
 'FELIXLÂNDIA',
 'LENÇÓIS',
 'MONTE ALTO',
 'FERNANDO FALCÃO',
 'RUBIATABA',
 'FRANCISCO SANTOS',
 'REDENÇÃO',
 'SÃO FERNANDO',
 'RIO DAS OSTRAS',
 'MARCAÇÃO',
 'SÃO MIGUEL DOS CAMPOS',
 'CASTRO',
 'CANITAR',
 'SÍTIO DO MATO',
 'CÓRREGO NOVO',
 'Sem informação',
 'RIO MARIA',
 'ENTRE RIOS DE MINAS',
 'ALFREDO CHAVES',
 'OLÍMPIA',
 'NOSSA SENHORA DO SOCORRO',
 'CORIBE',
 'PALMITAL',
 'SANANDUVA',
 'BERT

,Código da Emenda,Ano da Emenda,Tipo de Emenda,Código do Autor da Emenda,Nome do Autor da Emenda,Número da emenda,Localidade de aplicação do recurso,Código Município IBGE,Município,Código UF IBGE,...,Nome Ação,Código Plano Orçamentário,Nome Plano Orçamentário,Valor Empenhado,Valor Liquidado,Valor Pago,Valor Restos A Pagar Inscritos,Valor Restos A Pagar Cancelados,Valor Restos A Pagar Pagos,Código da Emenda_1
348,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SÃO PAULO - SP,3550308,SÃO PAULO,3500000,...,PROMOCAO DOS DIREITOS HUMANOS,0,PROMOCAO E DEFESA DOS DIREITOS HUMANOS - DESPE...,1499997.98,0.0,0.0,0.00,0.0,1499997.98,Sem informação
948,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SÃO PAULO - SP,3550308,SÃO PAULO,3500000,...,APOIO A MANUTENCAO DE UNIDADES DE SAUDE,0,APOIO A MANUTENCAO DE UNIDADES DE SAUDE - DESP...,100000.00,0.0,0.0,0.00,0.0,100000.00,Sem informação
1239,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,RIO DE JANEIRO - RJ,3304557,RIO DE JANEIRO,3300000,...,PROMOCAO E FOMENTO A CULTURA BRASILEIRA,0,PROMOCAO E FOMENTO A CULTURA BRASILEIRA - DESP...,1100900.41,150000.0,150000.0,0.00,60360.0,890540.41,Sem informação
1397,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SÃO PAULO - SP,3550308,SÃO PAULO,3500000,...,APOIO A INFRAESTRUTURA PARA A EDUCACAO BASICA,0,INFRAESTRUTURA PARA A EDUCACAO BASICA - DESPES...,1249500.00,0.0,0.0,0.00,0.0,1249500.00,Sem informação
1437,Sem informação,2014,Emenda Individual - Transferências com Finalid...,NaN,None,S/I,SÃO PAULO - SP,3550308,SÃO PAULO,3500000,...,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO PR...,0,ESTRUTURACAO DA REDE DE SERVICOS DE ATENCAO BA...,6468350.24,0.0,0.0,1190060.78,3092000.0,2186289.46,Sem informação


sp:343
rj:464


Rio de Janeiro tem mais emendas, 464, que São Paulo, 343

## 4) Na coluna `Código da Emenda`, substitua todos os valores `"Sem informação"` por `"Não disponível"`. Mostre a contagem de valores únicos da coluna antes e depois da substituição.

In [124]:
dados = dados.replace({'Código da Emenda' : {'Sem informação': 'Não disponível'}})
print(dados['Código da Emenda'].nunique())
print((dados['Código da Emenda'] == 'Não disponível').value_counts())

70731
Código da Emenda
False    70851
True     17808
Name: count, dtype: int64


## 5) Na coluna `Nome Função`, transforme todos os valores para que fiquem em letras maiúsculas. Confirme a alteração mostrando os valores únicos da coluna.

In [125]:
dados['Nome Função'] = dados['Nome Função'].str.upper()
dados['Nome Função'].unique().tolist()

['URBANISMO',
 'COMÉRCIO E SERVIÇOS',
 'SAÚDE',
 'EDUCAÇÃO',
 'AGRICULTURA',
 'ASSISTÊNCIA SOCIAL',
 'DESPORTO E LAZER',
 'DEFESA NACIONAL',
 'SANEAMENTO',
 'DIREITOS DA CIDADANIA',
 'GESTÃO AMBIENTAL',
 'CULTURA',
 'CIÊNCIA E TECNOLOGIA',
 'SEGURANÇA PÚBLICA',
 'ORGANIZAÇÃO AGRÁRIA',
 'COMUNICAÇÕES',
 'ADMINISTRAÇÃO',
 'HABITAÇÃO',
 'INDÚSTRIA',
 'TRABALHO',
 'MÚLTIPLO',
 'TRANSPORTE',
 'PREVIDÊNCIA SOCIAL',
 'ENCARGOS ESPECIAIS',
 'RELAÇÕES EXTERIORES',
 'ENERGIA',
 'ESSENCIAL À JUSTIÇA']

## 6) Verifique quantas linhas duplicadas existem no dataset. Em seguida, remova as duplicatas e mostre quantas linhas sobraram.

In [126]:
print(dados.duplicated().sum())
dados_uniq = dados.drop_duplicates()
print(dados_uniq.duplicated().sum())

63
0


## 7) Remova duplicatas considerando apenas as colunas `Código da Emenda` e `Município`. Quantas linhas sobraram? O resultado é diferente de quando removemos duplicatas do DataFrame inteiro? 

In [127]:
print(dados.duplicated(subset=['Código da Emenda', 'Município']).sum())
dados_uniq_municipio = dados.drop_duplicates(subset=['Código da Emenda', 'Município'])
len(dados_uniq_municipio)

12915


75744

Sobraram 75744 linhas, foram removidas 12915 enquanto foram removidas 63 duplicadas do DataFrame todo

## 8) Identifique quais colunas possuem valores nulos e quantos nulos cada uma tem. Em seguida, responda: qual é a coluna que possui maior número de valores nulos e quantos são?

In [128]:
dados.isnull().sum()

Código da Emenda                          0
Ano da Emenda                             0
Tipo de Emenda                            0
Código do Autor da Emenda             15961
Nome do Autor da Emenda               15961
Número da emenda                          0
Localidade de aplicação do recurso        0
Código Município IBGE                     0
Município                                 0
Código UF IBGE                            0
UF                                        0
Região                                    0
Código Função                             0
Nome Função                               0
Código Subfunção                          0
Nome Subfunção                            0
Código Programa                           0
Nome Programa                             0
Código Ação                               0
Nome Ação                                 0
Código Plano Orçamentário                 0
Nome Plano Orçamentário                   0
Valor Empenhado                 

O Código e o Nome do Autor da Emenda tem 15961 valores nulos sendo as unicas colunas com valores nulos

## 9) Preencha os valores nulos da coluna `codigo do autor` com **-1** e os valores nulos da coluna `nome do autor da emenda` com "desconhecido" e confirme que não restaram nulos nessa coluna.

In [129]:
dados['Código do Autor da Emenda'] = dados['Código do Autor da Emenda'].fillna(-1)
dados['Nome do Autor da Emenda'] = dados['Nome do Autor da Emenda'].fillna('Desconhecido')

dados.isnull().sum()

Código da Emenda                      0
Ano da Emenda                         0
Tipo de Emenda                        0
Código do Autor da Emenda             0
Nome do Autor da Emenda               0
Número da emenda                      0
Localidade de aplicação do recurso    0
Código Município IBGE                 0
Município                             0
Código UF IBGE                        0
UF                                    0
Região                                0
Código Função                         0
Nome Função                           0
Código Subfunção                      0
Nome Subfunção                        0
Código Programa                       0
Nome Programa                         0
Código Ação                           0
Nome Ação                             0
Código Plano Orçamentário             0
Nome Plano Orçamentário               0
Valor Empenhado                       0
Valor Liquidado                       0
Valor Pago                            0


## 10) Verifique o tipo da coluna `Ano da Emenda`. Converta-a para **string** (`str`). Mostre o tipo antes e depois da conversão.

In [130]:
print(dados['Ano da Emenda'].dtypes)
dados_ano_str = dados['Ano da Emenda'].astype(str)
print(dados_ano_str.dtypes)


int64
object


## 11) Converta a coluna `Valor Empenhado` para o tipo **float** (caso ainda não seja). Mostre o tipo após a conversão.

In [131]:
print(dados['Valor Empenhado'].dtypes)

float64


## 12) Calcule o **total** de `Valor Pago` por **Região**. Quais são as duas regiões que mais receberam recursos?

In [132]:
total_por_regiao = dados.groupby("Região")["Valor Pago"].sum().sort_values()

display(total_por_regiao)

Região
Exterior        2.572317e+07
Nacional        7.886148e+09
Centro-Oeste    1.037094e+10
Sul             1.648893e+10
Norte           2.035103e+10
Múltiplo        3.010563e+10
Sudeste         3.427210e+10
Nordeste        4.644963e+10
Name: Valor Pago, dtype: float64

Sudeste e Nordeste foram as regiões que mais receberam recursos com mais de 80 bilhões juntos

## 13) Calcule a **quantidade** de emendas por **Nome Função**. Mostre as 5 funções com mais emendas.

In [133]:
dados['Nome Função'].value_counts().head()

Nome Função
SAÚDE                 37068
URBANISMO              8663
EDUCAÇÃO               8108
AGRICULTURA            4940
ENCARGOS ESPECIAIS     4610
Name: count, dtype: int64

## 14) Crie uma nova coluna chamada `Saldo a Pagar`. Ela deve ser o resultado da subtração entre a coluna `Valor Empenhado` e a coluna `Valor Pago`. Mostre as 5 primeiras linhas apenas com essas três colunas.

In [134]:
dados['Saldo a Pagar'] = dados['Valor Empenhado'] - dados['Valor Pago']
dados[['Saldo a Pagar', 'Valor Pago', 'Valor Empenhado']].head()

,Saldo a Pagar,Valor Pago,Valor Empenhado
0,700000.00,0.0,700000.00
1,487500.00,0.0,487500.00
2,100000.00,0.0,100000.00
3,243750.00,0.0,243750.00
4,1658333.94,0.0,1658333.94


## 15) Elimine a(s) coluna(s) do dataframe que eestá totalmente vazia.

## 16) Para finalizar, salve o DataFrame transformado em um arquivo no formato JSON.